lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -453,6 +460,68 @@ void SPIRVRegularizeLLVMBase::lowerFuncPtr(Module *M) {
     lowerFuncPtr(I.first, I.second);
 }
 
+void SPIRVRegularizeLLVMBase::addKernelEntryPoint(Module *M) {
+  std::vector<Function *> Work;
+
+  // Get a list of all functions that have SPIR kernel calling conv
+  for (auto &F : *M) {
+    if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
+      continue;
+
+    Work.push_back(&F);
+  }
+  for (auto &F: Work) {
+
+    // for declarations just make them into SPIR functions.
+    if (F->empty()) {
+      F->setCallingConv(CallingConv::SPIR_FUNC);
+      continue;
+    }
+
+    // Otherwise add a wrapper around the function to act as an entry point.
+    FunctionType *FType = F->getFunctionType();
+    std::string wrap_name = "__spirv_entry_" + static_cast<std::string>(F->getName());
```

Comment:
I believe the idea should be not to create a new function, that is being used as an entry point, but instead to create a separate spir function, which is a copy of the kernel/entry point and then reuse this function to:
1. replace all function body of the kernel/entry point with a call to this function.
2. call this new function instead of a kernel/entry point, if it was called by another function.

I see that in the writer '__spirv_entry_' is being find and erased, but instead of it we can just have:
__kernel fun1() { // some code }
__kernel fun2() { call fun2 }

be transformed into

__kernel fun1_body() { // some code }
__kernel fun1() { call fun1_body() }    // we don't change a name of a kernel
__kernel fun2() { call fun1_body() }

Reply:
@MrSidims I'm not sure how this will work with linking though.
if I have two files with forward declerations I forsee issues

the same name needs to be in entry point and linkage attributes but pointing at two different functions.

if I have SPIR-V output from the above won't I get linkageattributes "fun2" for the function name, and any spir-v trying to link against this will be looking for fun1 linkage. (or am I missing something? happy to hear I am)

Reply:
file a.cl:
__kernel a() { //some code} 

file b.cl
__kernel a(); //fwd decl
__kernel b { // some code, a(), //more code }

Now how would b work out the renamed function a() for linkage here?

Reply:
@MrSidims any more info or do you think I'm wrong here.

Reply:
Sorry, I was on vacation back then and missed your answer. You are right.